manifold/balance.py

```diff
@@ -0,0 +1,153 @@
+        self.chain_id = chain_id
+
+        self.num_conns = num_conns
+        self.num_procs = num_procs
+
+        self.batch_size = batch_size
+        self.block_number = block_number
+
+        self.address = hex_to_bytes(BCHECKER_ADDRESSES[Network(chain_id)])
+
+    def aggregate(self) -> list[Balance]:
+        native_balances, erc20_balances = self._segment_balances()
+        balances: list[Balance] = []
+
+        if len(erc20_balances):
+            erc20 = MultiCall(
+                self.rpc_url,
+                [
+                    Call(
+                        balance.token_address,
+                        ERC20_BALANCE_SIGNATURE,
+                        (balance.token_address, balance.owner_address),
+                        input=(balance.owner_address,),
+                    )
+                    for balance in erc20_balances
+                ],
+                self.batch_size,
+                require_success=False,
+                chain_id=self.chain_id,
+                num_conns=self.num_conns,
+                num_procs=self.num_procs,
+                block_number=self.block_number,
+            )
+
+            balances += [
+                Balance(token_address, owner_address, value)
+                for (
+                    token_address,
+                    owner_address,
+                ), value in erc20.aggregate().items()
+            ]
+
+        if len(native_balances):
+            native = MultiCall(
+                self.rpc_url,
+                [
+                    Call(
+                        self.address,
+                        NATIVE_BALANCE_SIGNATURE,
+                        tuple(batch),
+                        input=(
+                            [ZERO_ADDRESS] * len(batch),
+                            [balance.owner_address for balance in batch],
+                        ),
+                    )
+                    for batch in batch(native_balances, self.batch_size)
+                ],
+                self.batch_size,
+                require_success=False,
+                chain_id=self.chain_id,
+                num_conns=self.num_conns,
+                num_procs=self.num_procs,
+                block_number=self.block_number,
+            )
+
+            balances += chain.from_iterable(
+                (
+                    Balance(balance.token_address, balance.owner_address, value)
+                    for balance, value in zip(balances, values)
+                )
+                for balances, values in native.aggregate().items()
+            )
+
+        return balances
+
+    def _segment_balances(self) -> tuple[list[BalanceRequest], list[BalanceRequest]]:
+        native_balances = []
+        erc20_balances = []
+
+        for balance in self.calls:
+            if balance.is_native():
+                native_balances.append(balance)
+            else:
+                erc20_balances.append(balance)
+
+        return native_balances, erc20_balances
```

Comment:
Why segment? Can't we use a single batch call with bchecker to get all balances?

Reply:
The underlying contract doesn't allow you to submit balance pairs. It lets you submit a list of addresses, and a list of tokens, then does a nested loop for each of them.

![image](https://github.com/Dedaub/manifold/assets/86597721/178cdbe1-972a-405a-9483-b760b6379281)

Reply:
Oh ok, I see.